### PR TITLE
[Bug fix] ttnn.topk: validate the preallocated output indices width

### LINKED ISSUE #51959: ttnn.topk: preallocated output indices dtype is not validated against the required index width
### Describe the bug

When `ttnn.topk` is given preallocated outputs via `output_tensor=(values, indices)`, the dtype of the indices tensor is never checked against the required index width. `compute_output_specs` simply returns the preallocated tensor config specs, bypassing the index-width decision made for calls where the output tensor isn't given (UINT16 only when the padded reduction dim fits in 16 bits *and* the values are not fp32), additionally `validate_on_program_cache_miss` only checks that the tensor is UINT16 or UINT32.

All checks below are from a single Blackhole p100a with `torch.manual_seed(0)`, `k=32`, `H=32`. The `idx_gather_mismatch` counts positions where `input[returned_index] != returned_value`.

**1. Preallocated UINT32 where UINT16 is required — the returned *values* are destroyed.** bf16 input at `W=64` requires UINT16:

```
prealloc UINT16 (correct)  ->  val_mismatch=   0/1024   idx_gather_mismatch=   0/1024
prealloc UINT32 (wider)    ->  val_mismatch=1018/1024   idx_gather_mismatch=1022/1024
```

**2. Preallocated UINT16 where UINT32 is required — indices are unrepresentable and wrap.** bf16 input at `W=131072` requires UINT32, because positions above 65535 do not fit in 16 bits:

```
prealloc UINT32 (correct)  ->  val_mismatch=0/1024   idx_gather_mismatch=  0/1024
prealloc UINT16 (narrower) ->  val_mismatch=0/1024   idx_gather_mismatch=502/1024
```

Roughly half the top-k positions of a random input fall above 65535 and wrap. With an input constructed so that *every* top-k element sits above position 65535, all 1024 returned indices equal the true index modulo 65536 — the first row expects `[65567, 65566, 65565, 65564]` and returns `[31, 30, 29, 28]`. Values are correct here; only the index tensor is wrong.

**3. fp32 + preallocated UINT16 is accepted and currently produces correct results, but likely only by accident.** 

### Steps to reproduce the issue

```python
import torch, ttnn

torch.manual_seed(0)
dev = ttnn.open_device(device_id=0)
k = 32

def run(W, idx_dt, label):
    shape, oshape = [1, 1, 32, W], [1, 1, 32, k]
    base = torch.randn(shape, dtype=torch.float32).bfloat16().float()
    x = ttnn.from_torch(base, ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=dev)
    z = torch.zeros(oshape, dtype=torch.float32)
    vt = ttnn.from_torch(z, ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=dev)
    it = ttnn.from_torch(z, idx_dt, layout=ttnn.TILE_LAYOUT, device=dev)
    v, i = ttnn.topk(x, k=k, dim=-1, largest=True, sorted=True, output_tensor=(vt, it))
    ttnn.synchronize_device(dev)
    gv, gi = ttnn.to_torch(v).float(), ttnn.to_torch(i).to(torch.int64)
    rv, _ = torch.topk(base, k, dim=-1, largest=True, sorted=True)
    vbad = int((gv != rv).sum())
    ibad = int((torch.gather(base, -1, gi.clamp(0, W - 1)) != gv).sum())
    print(f"{label:34s} val_mismatch={vbad:5d}/1024  idx_gather_mismatch={ibad:5d}/1024")

run(64,     ttnn.uint16, "W=64      UINT16 (correct)")
run(64,     ttnn.uint32, "W=64      UINT32 (defect 1)")
run(131072, ttnn.uint32, "W=131072  UINT32 (correct)")
run(131072, ttnn.uint16, "W=131072  UINT16 (defect 2)")
ttnn.close_device(dev)
```

Observed:

```
W=64      UINT16 (correct)   val_mismatch=    0/1024  idx_gather_mismatch=    0/1024
W=64      UINT32 (defect 1)  val_mismatch= 1018/1024  idx_gather_mismatch= 1022/1024
W=131072  UINT32 (correct)   val_mismatch=    0/1024  idx_gather_mismatch=    0/1024
W=131072  UINT16 (defect 2)  val_mismatch=    0/1024  idx_gather_mismatch=  502/1024
```

Expected: all four either correct, or the two mismatched pairings rejected with a clear error.

### Implication and priority

- **No model is affected.** No production caller passes `output_tensor=` to `ttnn.topk`; the parameter is currently exercised only by tests.
- **Existing coverage cannot catch this by construction.** `tests/ttnn/nightly/.../test_reduction_ops.py::_run_topk_with_preallocated` builds its preallocated buffers from the result of a prior non-preallocated call, so the dtypes always match what the op itself chose. A dtype mismatch is unreachable from that harness.
- Same root cause as #51921 (which covers the optional *input* `indices_tensor`) and as the review discussion on #50521: the index-width rule has no single owner. It is currently derived independently in `compute_output_specs`, in `validate_on_program_cache_miss`, in both program factories, and again caller-side in `models/common/sampling/tt_sampling.py::_select_topk_indices_dtype`.

Suggested direction, shared with #51921:

- Add a single `required_index_dtype(input_tensor, dim)` helper in `topk_utils`, and have `compute_output_specs`, `validate_on_program_cache_miss` and both program factories derive from it rather than re-deriving it.
- Validate the preallocated indices tensor against it with **exact equality**. Defect 1 shows that a wider tensor is broken too, so a "wide enough" relaxation is not safe.
- Apply the same check to the optional input `indices_tensor` (#51921).
- While there: the single-core factory's local `uint16_output` (`topk_single_core_program_factory.cpp:31`) uses `<` where `compute_output_specs` uses `<=`, and still feeds `fp32_dest_acc_en`. Deriving that flag from the index CB dtype instead would make "index CB is 32-bit" and "LLK uses INT32 index access" true by construction.

### PR DESCRIPTION
**Pull request by ign-sudha**
### PR Category

- Bug fix

### Summary

- Closes #51959

A preallocated output indices tensor sets the index width for the whole op, and nothing checked that width against what the input needs. A `uint16` indices buffer on an input needing 32 bits wrapped past 65535 and returned indices for the wrong columns, with correct values and no error.

Only defect 2 from the issue is left. Defect 1, a `uint32` buffer on a narrow input, was already resolved before #54168.

### Solution

```cpp
const DataType required_dtype = required_index_dtype(input_tensor, args.dim);
const bool indices_too_narrow = output_tensor1_dtype == DataType::UINT16 && required_dtype != DataType::UINT16;
TT_FATAL(
    !indices_too_narrow,
    "Preallocated indices tensor must be {} for this input, got: {}",
    required_dtype,
    output_tensor1_dtype);
```

Only too-narrow is rejected. A buffer wider than needed is correct and stays legal, so this is deliberately not `prealloc == required`. It also catches `float32` input, which needs 32-bit indices at any width because fp32 sorts with fp32 dest accumulation. 

### Tests

`test_topk_preallocated_indices_width` covers the five combinations, using `65536` for the wide cases. The passing rows gather the input by the returned indices and compare against the returned values, so a wrong index fails and not just a missing error.

### Notes for reviewers

- Behaviour change: a `uint16` preallocated indices buffer with a wide or `float32` input now raises instead of returning wrong indices. No caller in the repo preallocates topk outputs outside the tests.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:ign/sudha-51959-prealloc-output)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:ign/sudha-51959-prealloc-output)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:ign/sudha-51959-prealloc-output)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:ign/sudha-51959-prealloc-output)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:ign/sudha-51959-prealloc-output)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:ign/sudha-51959-prealloc-output)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:ign/sudha-51959-prealloc-output)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/reduction/topk/device/topk_device_operation.cpp [hunk 1]
```diff
@@ -280,6 +280,15 @@ void TopKDeviceOperation::validate_on_program_cache_miss(
                 output_tensor1_dtype == DataType::INT32,
             "Preallocated indices tensor must be UINT16, UINT32, or INT32 got: {}",
             output_tensor1_dtype);
+        // The preallocated indices tensor sets the index width for the whole op. A 16-bit one
+        // on an input that needs 32 bits wraps past 65535 and returns the wrong columns.
+        const DataType required_dtype = required_index_dtype(input_tensor, args.dim);
+        const bool indices_too_narrow = output_tensor1_dtype == DataType::UINT16 && required_dtype != DataType::UINT16;
```
Comment (iwroszTT) [inline ttnn/cpp/ttnn/operations/reduction/topk/device/topk_device_operation.cpp]: Let's improve `required_index_dtype`'s  API.

We know that `required_index_dtype` returns either UINT16 or UINT32, hence the condition for `indices_too_narrow` is correct.

Note that `DataType` enumerates other data types too, which might lead to errors in the future if `required_index_dtype ` is changed.
I believe the information we need from `required_index_dtype` is `is_uint32_index_required`, which allows for a simple true/false return value.

**Review by iwroszTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/reduction/topk/device/topk_device_operation.cpp (hunk 1 above)
Comment (ign-sudha) [reply ttnn/cpp/ttnn/operations/reduction/topk/device/topk_device_operation.cpp]: Updated in the latest commit. Please let me know if that looks okay.
